target/linux/ramips/mt7621/base-files/etc/uci-defaults/99_fwenv-store-ethaddr.sh

```diff
@@ -0,0 +1,15 @@
+[ ! -e /etc/fw_env.config ] && exit 0
+
+. /lib/functions/system.sh
+
+case "$(board_name)" in
+dna,valokuitu-plus-ex400)
+	ethaddr=$(/usr/sbin/fw_printenv -n ethaddr 2>/dev/null)
+	eth1addr=$(/usr/sbin/fw_printenv -n eth1addr 2>/dev/null)
+
+	[ -n "$ethaddr" ] && [ -z "$eth1addr" ] &&
+		/usr/sbin/fw_setenv eth1addr "$(macaddr_add $ethaddr 1)"
```

Comment:
I'm not very familiar with MediaTek / U-Boot. Could you tell my why do we need to set `eth1addr` variable? What is it used for?

Reply:
See the [my comment above where neheb asked the same question](https://github.com/openwrt/openwrt/pull/13798#discussion_r1712439206) . See also discussions between @dangowrt and @msandber. Vendor does not have eth1addr. It is a trick to let U-Boot set the MAC-address in the device tree using the ethernet1 alias. If we were able to use NVMEM on UBI already now, we could use ethaddr + 1 directly in DT, and we would not need this script.

Reply:
Oh, and those devices are not using MTK U-Boot, it was written by ourselves. Back then unfortunately MTK was not the MTK we have today so there was no U-Boot from MTK available: https://dev.iopsys.eu/fork/u-boot/-/commits/ex400/ (version on the device is some older commit of that branch)

This MR has been open for quite some time, I think if NVMEM on UBI does not arrive anytime soon, then maybe we can merge this device support PR as is, but right now it looks quite promising with NVMEM on UBI around the corner. Otherwise I would have suggested to merge the device support now and move to NVMEM later.

I am really excited about the better NVMEM support on UBI. NAND is not exactly a new technology and we have been storing stuff on UBI for several years, so it simplifies things a lot when we can finally specify things in device tree and not use so many shell-scripts extracting data.

Remind me, what is the current status of NVMEM on block/eMMC? Is it upstream now? I remember there were some objections to the patches, the last time I had checked.

Reply:
I believe NVMEM on UBI is already a thing. At least since February: https://github.com/openwrt/openwrt/commit/1794309bb5a442b5824de27a889f0855dcfe2198

Reply:
My suggestion is to:
1. Not set `eth1addr` OpenWrt custom variable
2. Use `target/linux/ramips/mt7621/base-files/etc/board.d/02_network` and `ucidef_set_interface_macaddr` for now. Maybe all you need is `wan_mac=$(macaddr_add $ethaddr 1)` or if not then some custom `ucidef_set_interface_macaddr "eth1" "$(macaddr_add $ethaddr 1)"`?
3. Switch to DT binding for UBI, NVMEM and `u-boot,env` layout when it's ready (it a month or something)

Reply:
I had something like this implemented previously: https://github.com/msandber/openwrt/commit/c1b9295f99cf1c7c0933342d01186ddcdcce01b5#diff-5669e3f33c1fdc220e9fe8ff7ade215fa65e754f41e5c2e9dda686ff592be74b

Back then there was an additional hickup of that relying on a new method `mtd_get_mac_ascii_ubi` in package/base-files/files/lib/functions/system.sh (https://github.com/msandber/openwrt/commit/cac8eb29488a6d7dadcdea4c6c27bcaa901e142f#diff-9d8640936a163ac354f846da8410231e303d9d63bb3cfb7f83685cd49b111cc3)

After that I went on to investigate different approaches to get the address outside user space and then finally landed with the current u-boot based ethernet address fixup with the `eth1addr` variable.

Before jumping through any more hoops I'd like to hear whether this is really the way to go now. Or should we wait for u-boot,env support on UBI volumes to enter mainline kernel?

Thanks.

Reply:
It hit me in my sleep that my approach has been ass-backwards and apologies for that. The eth0/lan address is already being set in DT so I can employ it to deduce eth1/wan address without digging anything from u-boot environment. I'll post a new version soonish.